File: LinkedList/Merge Two Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if(list1 == nullptr){return list2;}
        if(list2 == nullptr){return list1;}
        
        ListNode *sorted_head, *tmp;
        if(list1->val < list2->val){ 
            sorted_head = list1;
            list1 = list1->next;
        }
        else{
            sorted_head = list2;
            list2 = list2->next;
        }
        
        tmp = sorted_head;
        while(list1!=nullptr && list2!=nullptr){
            if(list1->val < list2->val){
                tmp->next = list1;
                list1 = list1->next;
            }
            else{
                tmp->next = list2;
                list2 = list2->next;
            }
            tmp = tmp->next;
        }
        
        while(list1!=nullptr){
            tmp->next = list1;
            list1 = list1->next;
            tmp = tmp->next;
        }
        
        while(list2!=nullptr){
            tmp->next = list2;
            list2 = list2->next;
            tmp = tmp->next;
        }
        
        return sorted_head;
    }
};

File: LinkedList/Merge Two Sorted Lists.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

impl Solution {
    pub fn merge_two_lists(mut list1: Option<Box<ListNode>>, mut list2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut dummy = ListNode::new(-1);
        let mut cur = &mut dummy;

        while let (Some(l1), Some(l2)) = (&list1, &list2) {
            if l1.val < l2.val {
                cur.next = list1.take();
                cur = cur.next.as_mut().unwrap();
                list1 = cur.next.take();
            } else {
                cur.next = list2.take();
                cur = cur.next.as_mut().unwrap();
                list2 = cur.next.take();
            }
        }
        cur.next = list1.or(list2);
        dummy.next
    }
}